### PR TITLE
[AptosFramework] prevent module republish

### PR DESCRIPTION
**Pull request by zekun000**

### CODE REVIEW

**Review by areshand — APPROVED** [review]

File: aptos-move/aptos-vm/src/aptos_vm_impl.rs [hunk 1]
```diff
@@ -302,13 +303,51 @@ impl AptosVMImpl {
...
+                Ok(module) => {
+                    // verify the sender matches the address
+                    if module.address() != &txn_data.sender {
+                        return Err(verification_error(
+                            StatusCode::MODULE_ADDRESS_DOES_NOT_MATCH_SENDER,
+                            IndexKind::AddressIdentifier,
+                            module.self_handle_idx().0,
+                        )
+                        .finish(Location::Undefined));
+                    }
```
Comment (davidiw) [inline aptos-move/aptos-vm/src/aptos_vm_impl.rs:322]: this is also checked upstream, is there a point in checking it again?

**Review by davidiw — APPROVED** [review]
looks good except duplicate code: https://github.com/move-language/move/blob/11bed4ecf3032bcfab0d03b88455ebfe54c207fb/language/move-vm/runtime/src/runtime.rs#L98

File: aptos-move/aptos-vm/src/aptos_vm_impl.rs (hunk 1 above)
Comment (zekun000) [reply aptos-move/aptos-vm/src/aptos_vm_impl.rs:322]: this is validation, so we can fail before execution

**Comment by Manahip** [comment]
I wish there was a mechanism to properly constrain module upgrades. On Solana where contracts are easily upgraded.. devs bump into all sorts of issues because some upstream contract can change their interface however they like.

**Pull request closed without merging**